lib/streamlit/runtime/fragment.py

```diff
@@ -0,0 +1,237 @@
+Fragment = Callable[[], Any]
+
+
+class FragmentStorage(Protocol):
+    """A key-value store for Fragments. Used to implement the @st.experimental_fragment
+    decorator.
+
+    We intentionally define this as its own protocol despite how generic it appears to
+    be at first glance. The reason why is that, in any case where fragments aren't just
+    stored as Python closures in memory, storing and retrieving Fragments will generally
+    involve serializing and deserializing function bytecode, which is a tricky aspect
+    to implementing FragmentStorages that won't generally appear with our other *Storage
+    protocols.
+    """
+
+    @abstractmethod
+    def get(self, key: str) -> Fragment:
+        """Returns the stored fragment for the given key."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def set(self, key: str, value: Fragment) -> None:
+        """Saves a fragment under the given key."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def delete(self, key: str) -> None:
+        """Delete the fragment corresponding to the given key."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def clear(self) -> None:
+        """Remove all fragments saved in this FragmentStorage."""
+        raise NotImplementedError
+
+
+# NOTE: Ideally, we'd like to add a MemoryFragmentStorageStatProvider implementation to
+# keep track of memory usage due to fragments, but doing something like this ends up
+# being difficult in practice as the memory usage of a closure is hard to measure (the
+# vendored implementation of pympler.asizeof that we use elsewhere is unable to measure
+# the size of a function).
+class MemoryFragmentStorage(FragmentStorage):
+    """A simple, memory-backed implementation of FragmentStorage.
+
+    MemoryFragmentStorage is just a wrapper around a plain Python dict that complies with
+    the FragmentStorage protocol.
+    """
+
+    def __init__(self):
+        self._fragments: dict[str, Fragment] = {}
+
+    def get(self, key: str) -> Fragment:
+        return self._fragments[key]
+
+    def set(self, key: str, value: Fragment) -> None:
+        self._fragments[key] = value
+
+    def delete(self, key: str) -> None:
+        del self._fragments[key]
+
+    def clear(self) -> None:
+        self._fragments.clear()
+
+
+@overload
+def fragment(
+    func: F,
+    *,
+    run_every: int | float | timedelta | str | None = None,
+) -> F:
+    ...
+
+
+# Support being able to pass parameters to this decorator (that is, being able to write
+# `@fragment(run_every=5.0)`).
+@overload
+def fragment(
+    func: None = None,
+    *,
+    run_every: int | float | timedelta | str | None = None,
+) -> Callable[[F], F]:
+    ...
+
+
+@gather_metrics("experimental_fragment")
+def fragment(
+    func: F | None = None,
+    *,
+    run_every: int | float | timedelta | str | None = None,
+) -> Callable[[F], F] | F:
+    """Allow a function to be run independently of the full script.
```

Comment:
btw @sfc-gh-jcarroll @sfc-gh-dmatthews, I wrote this docstring just to have something reasonable for now, but I'm assuming we'll want to tweak/edit it before this PR gets merged

Reply:
This looks fairly reasonable to me, I'll happily defer to you and Debbie on any tweaks or suggestions of bigger things to include.

In case it's a useful reference, here's the (significantly more verbose) [draft doc string](https://www.notion.so/snowflake-corp/Product-Spec-eeb691a8642d41e8969f1c82f8ac2370?pvs=4#9a4e84d2f77d400b92593ddb0fb49b60) I put together for the spec. I think the main interesting questions from my version are:

- For the first line, is it useful to include "...turn a function into a [Streamlit] fragment..." along with the description of what it does? I think the answer partly depends on how much we want to lean into that as a noun we use elsewhere.
- Whether we want to include any of the last 4 paragraphs of my version in the docstring, or just talk about those in the docs page.

Reply:
@vdonato I'll render the new docs page to check it out. Do you prefer a follow-up PR with any docstring tweaks, or a commit directly to this branch (if it hasn't merged already).

Reply:
Whoops, I totally forgot that a draft docstring was included in the spec 🙈. Happy to use either one or some combination of the two, but will also defer to @sfc-gh-dmatthews on this.

This PR is likely to be merged Thursday or Friday depending on how quickly QA is able to wrap up some testing passes on the feature, so committing directly to this branch should be ok if it isn't already checked in.

Reply:
Looks like QA wrapped up earlier than expected, so I'm going to go ahead and click merge on this PR now / any docs tweaks can be made in a PR to develop 🎉